pyppeteer/network_manager.py

```diff
@@ -497,160 +488,137 @@ def failure(self) -> Optional[Dict]:
         - ``timedout``: An operation timed out.
         - ``failed``: A generic failure occurred.
         """
-        if self._url.startswith('data:'):
+        if not self._actionable_request:
             return
-        errorReason = errorReasons[errorCode]
+
+        errorReason = self._errorReasons.get(errorCode)
         if not errorReason:
-            raise NetworkError('Unknown error code: {}'.format(errorCode))
-        if not self._allowInterception:
-            raise NetworkError('Request interception is not enabled.')
-        if self._interceptionHandled:
-            raise NetworkError('Request is already handled.')
+            raise NetworkError(f'Unknown error code: {errorCode}')
+
         self._interceptionHandled = True
         try:
             await self._client.send(
                 'Fetch.failRequest', {'requestId': self._interceptionId, 'errorReason': errorReason}
             )
         except Exception as e:
+            # In certain cases, protocol will return error if the request was already canceled
+            # or the page was closed. We should tolerate these errors.
             debugError(logger, e)
 
+    @property
+    def _actionable_request(self) -> bool:
+        """
+        Checks if we can abort/continue/respond to request.
+        :return: True if we can, False if we can't
+        """
+        # Mocking responses for dataURL requests is not currently supported.
+        if self._url.startswith('data:'):
+            return False
+        if not self._allowInterception:
+            raise ValueError('Request Interception is not enabled!')
+        if self._interceptionHandled:
+            raise ValueError('Request is already handled')
+        return True
+
 
-errorReasons = {
-    'aborted': 'Aborted',
-    'accessdenied': 'AccessDenied',
-    'addressunreachable': 'AddressUnreachable',
-    'blockedbyclient': 'BlockedByClient',
-    'blockedbyresponse': 'BlockedByResponse',
-    'connectionaborted': 'ConnectionAborted',
-    'connectionclosed': 'ConnectionClosed',
-    'connectionfailed': 'ConnectionFailed',
-    'connectionrefused': 'ConnectionRefused',
-    'connectionreset': 'ConnectionReset',
-    'internetdisconnected': 'InternetDisconnected',
-    'namenotresolved': 'NameNotResolved',
-    'timedout': 'TimedOut',
-    'failed': 'Failed',
-}
-
-
-class Response(object):
+class Response:
     """Response class represents responses which are received by ``Page``."""
 
-    def __init__(
-        self,
-        client: CDPSession,
-        request: Request,
-        status: int,
-        headers: Dict[str, str],
-        fromDiskCache: bool,
-        fromServiceWorker: bool,
-        securityDetails: Dict = None,
-        remoteIpAddress: str = '',
-        remotePort: str = '',
-        statusText: str = '',
-    ) -> None:
+    def __init__(self, client: CDPSession, request: 'Request', responsePayload):
```

Comment:
We should probably introduce TypedDict for responsePayload. They have field definitions in js code - it's quite long though :D

Reply:
I agree. I was looking at puppeteer and it looks like they generate definitions dynamically - I still haven't gotten around to running puppeteer from source and I think we can wait on this.